aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
 #else // AT_MKLDNN_EBABLED
 
 #include <ATen/mkldnn/Runtime.h>
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
 
 using namespace mkldnn;
 
 namespace at { namespace native {
 
-constexpr int input_batch_size_dim = 0;  // also grad_input
-constexpr int input_channels_dim = 1;
-constexpr int output_batch_size_dim = 0;  // also grad_output
-constexpr int output_channels_dim = 1;
-constexpr int weight_output_channels_dim = 0;
-constexpr int weight_input_channels_dim = 1;
+Tensor _ideep_conv2d(
+    const Tensor& input, const Tensor& weight, const Tensor& bias,
+    at::IntArrayRef padding, at::IntArrayRef stride, at::IntArrayRef dilation, int64_t groups) {
 
-// Often written as 2 + max_dim (extra dims for batch size and channels)
-constexpr int max_dim = 3;
+  auto stride_vec = expand_param_if_needed(stride, 2);
+  auto padding_vec = expand_param_if_needed(padding, 2);
+  auto dilation_vec = expand_param_if_needed(dilation, 2);
 
-static std::vector<int64_t> conv_output_size(
-    IntArrayRef input_size, IntArrayRef weight_size,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto dim = input_size.size();
-  std::vector<int64_t> output_size(dim);
-  output_size[0] = input_size[input_batch_size_dim];
-  output_size[1] = weight_size[weight_output_channels_dim];
-  for (size_t d = 2; d < dim; ++d) {
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+  const std::vector<int64_t> output_sizes = conv_output_size(
+    input.sizes(), weight.sizes(), padding, stride, dilation);
+  ideep::scale_t dummy_scale;
+  ideep::tensor y;
+  if (bias.defined()) {
+    const ideep::tensor& b = itensor_from_mkldnn(bias);
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      b,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {dilation_vec.cbegin(), dilation_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      groups,
+      dummy_scale,
+      dummy_scale,
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward_inference);
```

Comment:
Suggest to use `ideep::prop_kind::forward` instead since the same function would be used in both training and inference.

Reply:
Good point. will change it.
Actually I'm curious what's the perf difference between forward_inference and forward_training for ideep conv?

Reply:
@bddppq So far, the only difference lies in Winograd conv. For inference, weight transformation is only done once and transformed weight can be cached in MKL-DNN opaque format (the same way as other blocking formats). For training, weight is transformed all the time.